### PR TITLE
 WPS465: Forbid to use `&` and `|` with booleans #1657 

### LINKED ISSUE #1657: Forbid to use `&` and `|` with booleans
# Rule request

## Thesis

Yesterday I saw this code: 

```python
my_result = (first > 0 & second != 0) | other is None
```

And I was 😮 

So, we need to check all operations around `&` and `|` and detect any booleans there:
- Explicit `True` and `False`
- Compares with `is` or `==` or `!=`
- Compares with `>`, `>=`, etc
- Logical operations `not`, `or`, `and`

If we can detect at least one of these operations, then we should raise a violation. 
Let's call it `BinaryAndBooleanMixupViolation`.

Note: other binary operations are still fine!
So, these lines should not raise:
- `first & second`
- `first | 10`
- etc

## Reasoning

This case indicates that a person confused `&` with `and` and `|` with `or`.
This can be the case if a person is comming from another language.

### PR DESCRIPTION
**Pull request by thammaji**
# I have made things!

WPS465: It is forbidden to use '|' and '&' (bitwise operators) with booleans. For example:

```
(x >= y) & True
(first > 0 & second != 0) | (other is None)
```

But usage of bitwise operators numbers, variables or with other binary operators is fine:

```
first & second
(5 + 324 + 35) & y
```
Usage of bitwise operators with explicit `True` or `False`, boolean operators (`and` `or`), unary operators (`not`), and comparisons (`<` `!=`) will be forbidden.

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #1657

### CODE REVIEW

File: tests/test_checker/test_noqa.py [hunk 1]
```diff
@@ -237,6 +237,7 @@
     'WPS462': 1,
     'WPS463': 1,
     'WPS464': 0,  # logically unacceptable.
+    'WPS465': 0,  # logically unacceptable.
```
Comment (sobolevn) [inline tests/test_checker/test_noqa.py:240]: It is perfectly acceptable and should be present in `fixtures/noqa.py`

File: tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from wemake_python_styleguide.violations.best_practices import (
+    BitwiseAndBooleanMixupViolation,
+)
+from wemake_python_styleguide.visitors.ast.operators import BitwiseOpVisitor
+
+# add more cases for more usages of mixup
+
+
+@pytest.mark.parametrize('expression', [
+    'True | False',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:12]: Let's also test `True | True`, `False & False`

File: tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    '(x >= y) & True',
+    'first | (10 == second)',
+    '(11 != first) | (not False)',
+    '(1 or first) & (second and first)',
+    '(first > 0 & second != 0) | (other is None)',
+    '(5234 + 2345) & True',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:18]: What about `None`? Should it be counted?

File: tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+])
+def test_bitwise_boolean_mixup(
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:20]: Can we have some tests without `()`? Or will the operator order break?

File: wemake_python_styleguide/violations/best_practices.py [hunk 5]
```diff
@@ -2504,3 +2506,36 @@ class EmptyCommentViolation(TokenizeViolation):
...
+
+    Solution:
+        Change bitwise operator to boolean operators.
+
+    Example::
+
+        # Correct:
+
+        first | 10
+        result = ((first > 0) && False)
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2530]: Just the first one is enough, please remove this line.

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 6]
```diff
@@ -268,3 +269,35 @@ def visit_NamedExpr(
...
+
+    def visit_BinOp(self, node: ast.BinOp) -> None:
+        """
+        Finds bad usage of bitwise operation with binary operation.
+
+        Raises:
+            BitwiseAndBooleanMixupViolation
+
+        """
+        self._check_BitOp(node)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:286]: Please, name it logically, not structurely. 

For example: `_check_logical_bitwise_operator`

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 7]
```diff
@@ -268,3 +269,35 @@ def visit_NamedExpr(
...
+        self.generic_visit(node)
+
+    def _check_BitOp(self, node: ast.BinOp) -> None:
+        if not isinstance(node.op, (ast.BitOr, ast.BitAnd)):
+            return
+
+        if (self._check_sides(node.left) or self._check_sides(node.right)):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:293]: you don't need `()` here.

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 8]
```diff
@@ -268,3 +269,35 @@ def visit_NamedExpr(
...
+            self.add_violation(BitwiseAndBooleanMixupViolation(node))
+
+    # checks either side of the Bitwise operation invalid usage
+    def _check_sides(self, node) -> bool:
+        invalid = False
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:298]: We can refactor this into:

`return isinstance(node, self._invalid_nodes)`

Where `self._invalid_nodes` is a class-level constant with all the nodes we need. Check other visitor for examples, search `AnyNodes` type.

**Review by sobolevn — CHANGES_REQUESTED** [review]
Awesome work! There's a bit of polishing required, but looks like we are almost ready!

File: tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py (hunk 3 above)
Comment (thammaji) [reply tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:18]: I think it should be counted as a violation (python will  throw an error when NoneType is used with Bitwise operators anyways).  Added `None` test case.

File: tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py (hunk 4 above)
Comment (thammaji) [reply tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:20]: The operator order will break, `1 or first & second and first` will not raise a violation, as the left and right of the BitwiseOp are just Name variables, and I don't think there is any way to traverse the AST from Name variables. This means that there will be some false negatives, but I don't think that there's a better way.

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_operators/test_bitwise_boolean_mixup.py:20]: Ok 👍

File: wemake_python_styleguide/visitors/ast/operators.py (hunk 8 above)
Comment (thammaji) [reply wemake_python_styleguide/visitors/ast/operators.py]: Thanks for this suggestion, I was feeling that my code there was a little inelegant.

File: tests/fixtures/noqa/noqa.py [hunk 9]
```diff
@@ -765,3 +765,5 @@ def many_raises_function(parameter):  # noqa: WPS238
 
 def get_item():  # noqa: WPS463
     return  # noqa: WPS324
+
+bad_bitwise = True | True # noqa: WPS465
```
Comment (sobolevn) [inline tests/fixtures/noqa/noqa.py:769]: Should end with a new line

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 10]
```diff
@@ -283,21 +290,16 @@ def visit_BinOp(self, node: ast.BinOp) -> None:
...
+    def _check_logical_bitwise_operator(self, node: ast.BinOp) -> None:
         if not isinstance(node.op, (ast.BitOr, ast.BitAnd)):
             return
 
-        if (self._check_sides(node.left) or self._check_sides(node.right)):
+        if self._check_sides(node.left) or self._check_sides(node.right):
             self.add_violation(BitwiseAndBooleanMixupViolation(node))
 
     # checks either side of the Bitwise operation invalid usage
     def _check_sides(self, node) -> bool:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:304]: Let's better name it `_is_bool_like`. What do you think?

Comment (thammaji) [reply wemake_python_styleguide/visitors/ast/operators.py:304]: Sure, that's a would be more descriptive of what the function actually does.

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 11]
```diff
@@ -268,3 +269,37 @@ def visit_NamedExpr(
...
+
+    def _check_logical_bitwise_operator(self, node: ast.BinOp) -> None:
+        if not isinstance(node.op, (ast.BitOr, ast.BitAnd)):
+            return
+
+        if self._is_bool_like(node.left) or self._is_bool_like(node.right):
+            self.add_violation(BitwiseAndBooleanMixupViolation(node))
+
+    # checks either side of the Bitwise operation invalid usage
+    def _is_bool_like(self, node) -> bool:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:304]:
```suggestion
    def _is_bool_like(self, node) -> bool:
    def _is_bool_like(self, node: ast.expr) -> bool:
```

**Review by sobolevn — APPROVED** [review]
Awesome, thank you! Very cool feature 👍

**Comment by thammaji** [comment]
I had a good experience working on it! Thanks for being so friendly and welcoming!

**Comment by sobolevn** [comment]
Thanks 😊

**Pull request merged**